File: hello-dart/lib/server/memory/memory_pool_manager.dart
// Memory pool manager for efficient buffer reuse.
// Pre-allocates buffers to minimize allocation overhead.
// Implemented as a singleton to ensure a single shared pool across all streams.
// Matches C++ MemoryPoolManager and Java MemoryPoolManager functionality.

import 'dart:typed_data';
import '../../src/logger.dart';

/// Memory pool manager singleton.
class MemoryPoolManager {
  static MemoryPoolManager? _instance;
  final int bufferSize;
  final int poolSize;
  final List<List<int>> _availableBuffers;
  int _totalBuffers = 0;

  /// Get the singleton instance of MemoryPoolManager.
  static MemoryPoolManager getInstance(
      [int bufferSize = 65536, int poolSize = 100]) {
    _instance ??= MemoryPoolManager._internal(bufferSize, poolSize);
    return _instance!;
  }

  /// Private constructor for singleton pattern.
  MemoryPoolManager._internal(this.bufferSize, this.poolSize)
      : _availableBuffers = [] {
    // Pre-allocate buffers
    for (int i = 0; i < poolSize; i++) {
      Uint8List buffer = Uint8List(bufferSize);
      _availableBuffers.add(buffer);
      _totalBuffers++;
    }

    Logger.info(
        'MemoryPoolManager initialized with $poolSize buffers of $bufferSize bytes');
  }

  /// Acquire a buffer from the pool.
  /// If pool is exhausted, allocates a new buffer dynamically.
  List<int> acquireBuffer() {
    if (_availableBuffers.isNotEmpty) {
      var buffer = _availableBuffers.removeAt(0);
      Logger.debug(
          'Acquired buffer from pool (${_availableBuffers.length} remaining)');
      return buffer;
    } else {
      // Pool exhausted, allocate new buffer
      var buffer = Uint8List(bufferSize);
      _totalBuffers++;
      Logger.debug(
          'Pool exhausted, allocated new buffer (total: $_totalBuffers)');
      return buffer;
    }
  }

  /// Release a buffer back to the pool.
  void releaseBuffer(List<int> buffer) {
    if (buffer.length != bufferSize) {
      Logger.warn(
          'Buffer size mismatch: expected $bufferSize, got ${buffer.length}');
      return;
    }

    // Clear buffer before returning to pool
    for (int i = 0; i < buffer.length; i++) {
      buffer[i] = 0;
    }

    // Only return to pool if we haven't exceeded pool size
    if (_availableBuffers.length < poolSize) {
      _availableBuffers.add(buffer);
    }

    Logger.debug(
        'Released buffer to pool (${_availableBuffers.length} available)');
  }

  /// Get the number of available buffers in the pool.
  int getAvailableBuffers() => _availableBuffers.length;

  /// Get the total number of buffers (available + in-use).
  int getTotalBuffers() => poolSize;

  /// Get the buffer size.
  int getBufferSize() => bufferSize;
}

File: hello-swift/Sources/Server/Memory/MemoryPoolManager.swift
//
//  MemoryPoolManager.swift
//  Audio Stream Server
//
//  Memory pool manager for efficient buffer reuse.
//  Pre-allocates buffers to minimize allocation overhead.
//  Implemented as a singleton to ensure a single shared pool across all streams.
//  Matches C++ MemoryPoolManager and Java MemoryPoolManager functionality.
//

import Foundation
import AudioStreamCommon

/// Memory pool manager singleton.
class MemoryPoolManager {
    private static var instance: MemoryPoolManager?
    private let bufferSize: Int
    private let poolSize: Int
    private var availableBuffers: [Data]
    private var totalBuffers: Int
    private let lock = NSLock()

    /// Get the singleton instance of MemoryPoolManager.
    static func getInstance(bufferSize: Int = 65536, poolSize: Int = 100) -> MemoryPoolManager {
        if let instance = instance {
            return instance
        }

        let newInstance = MemoryPoolManager(bufferSize: bufferSize, poolSize: poolSize)
        instance = newInstance
        return newInstance
    }

    /// Private constructor for singleton pattern.
    private init(bufferSize: Int, poolSize: Int) {
        self.bufferSize = bufferSize
        self.poolSize = poolSize
        self.availableBuffers = []
        self.totalBuffers = 0

        // Pre-allocate buffers
        for _ in 0..<poolSize {
            let buffer = Data(repeating: 0, count: bufferSize)
            availableBuffers.append(buffer)
            totalBuffers += 1
        }

        Logger.info("MemoryPoolManager initialized with \(poolSize) buffers of \(bufferSize) bytes")
    }

    /// Acquire a buffer from the pool.
    /// If pool is exhausted, allocates a new buffer dynamically.
    func acquireBuffer() -> Data {
        lock.lock()
        defer { lock.unlock() }

        if !availableBuffers.isEmpty {
            let buffer = availableBuffers.removeFirst()
            Logger.debug("Acquired buffer from pool (\(availableBuffers.count) remaining)")
            return buffer
        } else {
            // Pool exhausted, allocate new buffer
            let buffer = Data(repeating: 0, count: bufferSize)
            totalBuffers += 1
            Logger.debug("Pool exhausted, allocated new buffer (total: \(totalBuffers))")
            return buffer
        }
    }

    /// Release a buffer back to the pool.
    func releaseBuffer(_ buffer: Data) {
        if buffer.count != bufferSize {
            Logger.warn("Buffer size mismatch: expected \(bufferSize), got \(buffer.count)")
            return
        }

        lock.lock()
        defer { lock.unlock() }

        // Clear buffer before returning to pool
        let cleared = Data(repeating: 0, count: bufferSize)

        // Only return to pool if we haven't exceeded pool size
        if availableBuffers.count < poolSize {
            availableBuffers.append(cleared)
        }

        Logger.debug("Released buffer to pool (\(availableBuffers.count) available)")
    }

    /// Get the number of available buffers in the pool.
    func getAvailableBuffers() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return availableBuffers.count
    }

    /// Get the total number of buffers (available + in-use).
    func getTotalBuffers() -> Int {
        return poolSize
    }

    /// Get the buffer size.
    func getBufferSize() -> Int {
        return bufferSize
    }
}
